libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+#include <foundation.h>
+#include <foundation-auto.h>
+
+#include "foundation-private.h"
+#include "foundation-java-private.h"
+
+static MCJavaType MCJavaMapTypeCodeSubstring(MCStringRef p_type_code, MCRange p_range)
+{
+    if (MCStringSubstringIsEqualToCString(p_type_code, p_range, "[",
+                                          kMCStringOptionCompareExact))
+        return kMCJavaTypeArray;
+    
+    for (uindex_t i = 0; i < sizeof(type_map) / sizeof(type_map[0]); i++)
+    {
+        if (MCStringSubstringIsEqualToCString(p_type_code, p_range, type_map[i] . name, kMCStringOptionCompareExact))
+        {
+            return type_map[i] . type;
+        }
+    }
+    
```

Comment:
FYI, you can do some clever stuff with the STL to avoid the `sizeof()` ratio dance, e.g.:

```c++
#include <algorithm>

...

auto t_mapping = std::find(std::begin(type_map), std::end(type_map),
    [&](const java_type_map& p_entry) {
        return MCStringSubstringIsEqualToCString(p_type_code, p_range,
            p_entry.name, kMCStringOptionCompareExact);
    });
if (t_mapping != std::end(type_map))
    return t_mapping->type;
else
    return kMCJavaTypeObject;
```